### PR TITLE
sha512: use openssl implementation if available

### PR DESCRIPTION
**Pull request by yuryu**
OpenSSL provides an optimized implementation of SHA512. This patch
changes flb_sha512_*() functions to use them if available. Otherwise
they continue to use the current implementation.

Signed-off-by: Emma Haruka Iwao <yuryu@google.com>

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [N/A] Example configuration file for the change
- [X] Debug log output from testing the change

- [X] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

```
valgrind bin/fluent-bit -c fin
==133849== Memcheck, a memory error detector
==133849== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==133849== Using Valgrind-3.17.0 and LibVEX; rerun with -h for copyright info
==133849== Command: bin/fluent-bit -c fin
==133849==
Fluent Bit v1.9.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/11/09 23:21:55] [ info] Configuration:
[2021/11/09 23:21:55] [ info]  flush time     | 5.000000 seconds
[2021/11/09 23:21:55] [ info]  grace          | 5 seconds
[2021/11/09 23:21:55] [ info]  daemon         | 0
[2021/11/09 23:21:55] [ info] ___________
[2021/11/09 23:21:55] [ info]  inputs:
[2021/11/09 23:21:55] [ info]      cpu
[2021/11/09 23:21:55] [ info] ___________
[2021/11/09 23:21:55] [ info]  filters:
[2021/11/09 23:21:55] [ info] ___________
[2021/11/09 23:21:55] [ info]  outputs:
[2021/11/09 23:21:55] [ info]      forward.0
[2021/11/09 23:21:55] [ info] ___________
[2021/11/09 23:21:55] [ info]  collectors:
[2021/11/09 23:21:55] [ info] [engine] started (pid=133849)
[2021/11/09 23:21:55] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2021/11/09 23:21:55] [debug] [storage] [cio stream] new stream registered: cpu.0
[2021/11/09 23:21:55] [ info] [storage] version=1.1.5, initializing...
[2021/11/09 23:21:55] [ info] [storage] in-memory
[2021/11/09 23:21:55] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/11/09 23:21:55] [ info] [cmetrics] version=0.2.2
[2021/11/09 23:21:55] [debug] [forward:forward.0] created event channels: read=19 write=20
[2021/11/09 23:21:55] [debug] [router] match rule cpu.0:forward.0
[2021/11/09 23:21:55] [ info] [sp] stream processor started
[2021/11/09 23:21:59] [debug] [task] created task=0x4fe4bd0 id=0 OK
[2021/11/09 23:21:59] [debug] [output:forward:forward.0] request 4420 bytes to flush
[2021/11/09 23:22:00] [debug] [output:forward:forward.0] protocol: received HELO
[2021/11/09 23:22:00] [debug] [output:forward:forward.0] PING sent: ret=294 bytes sent=294
[2021/11/09 23:22:00] [debug] [output:forward:forward.0] handshake status = 0
[2021/11/09 23:22:00] [debug] [upstream] KA connection #25 to 127.0.0.1:24224 is now available
[2021/11/09 23:22:00] [debug] [out coro] cb_destroy coro_id=0
[2021/11/09 23:22:00] [debug] [task] destroy task=0x4fe4bd0 (task_id=0)
[2021/11/09 23:22:04] [debug] [task] created task=0x5083840 id=0 OK
[2021/11/09 23:22:04] [debug] [output:forward:forward.0] request 6630 bytes to flush
[2021/11/09 23:22:04] [debug] [upstream] KA connection #25 to 127.0.0.1:24224 has been assigned (recycled)
[2021/11/09 23:22:04] [debug] [upstream] KA connection #25 to 127.0.0.1:24224 is now available
[2021/11/09 23:22:04] [debug] [out coro] cb_destroy coro_id=1
[2021/11/09 23:22:04] [debug] [task] destroy task=0x5083840 (task_id=0)
[2021/11/09 23:22:09] [debug] [task] created task=0x50ee010 id=0 OK
[2021/11/09 23:22:09] [debug] [output:forward:forward.0] request 4420 bytes to flush
[2021/11/09 23:22:09] [debug] [upstream] KA connection #25 to 127.0.0.1:24224 has been assigned (recycled)
[2021/11/09 23:22:09] [debug] [upstream] KA connection #25 to 127.0.0.1:24224 is now available
[2021/11/09 23:22:09] [debug] [out coro] cb_destroy coro_id=2
[2021/11/09 23:22:09] [debug] [task] destroy task=0x50ee010 (task_id=0)
^C[2021/11/09 23:22:10] [engine] caught signal (SIGINT)
[2021/11/09 23:22:10] [ info] [input] pausing cpu.0
[2021/11/09 23:22:10] [debug] [task] created task=0x513f030 id=0 OK
[2021/11/09 23:22:10] [debug] [output:forward:forward.0] request 1105 bytes to flush
[2021/11/09 23:22:10] [debug] [upstream] KA connection #25 to 127.0.0.1:24224 has been assigned (recycled)
[2021/11/09 23:22:10] [debug] [upstream] KA connection #25 to 127.0.0.1:24224 is now available
[2021/11/09 23:22:10] [ warn] [engine] service will stop in 5 seconds
[2021/11/09 23:22:10] [debug] [out coro] cb_destroy coro_id=3
[2021/11/09 23:22:10] [debug] [task] destroy task=0x513f030 (task_id=0)
[2021/11/09 23:22:14] [ info] [engine] service stopped
==133849==
==133849== HEAP SUMMARY:
==133849==     in use at exit: 0 bytes in 0 blocks
==133849==   total heap usage: 1,342 allocs, 1,342 frees, 2,128,725 bytes allocated
==133849==
==133849== All heap blocks were freed -- no leaks are possible
==133849==
==133849== For lists of detected and suppressed errors, rerun with: -s
==133849== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

with the following config
```
[SERVICE]
    Flush      5
    Daemon     off
    Log_Level  debug
[INPUT]
    Name       cpu
    Tag        cpu_usage
[OUTPUT]
    Name          forward
    Match         *
    Host          127.0.0.1
    Port          24224
    Shared_Key    secret
    Self_Hostname flb.local
    tls           off
    tls.verify    off
```

Tested against fluentd 1.14.2 configured as
```
<source>
  @type         forward
  <security>
    self_hostname myserver.local
    shared_key    secret
  </security>
  bind 0.0.0.0
  port 24224
</source>
<match **>
 @type stdout
</match>
```

**Documentation**

- [N/A] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Review by leonardo-albertovich — COMMENTED** [review]
It seems like according to the documentation these functions could fail although it doesn't specify under which circumstances and our own implementation does not have a return code which makes me wonder if we should just ignore those or add a new field to the structure that can hold a numeric error code and an error getter with a compatible ABI in case the client code wants to be sure everything went well.

Since all checks passed I assume no existing code accesses the `flb_sha512` structure fields directly so I think this patch is good to go once we determine if we do want to add that error code feature or not.

**Comment by yuryu** [comment]
We might not want to depend on the current implementation but as of today ([link](https://github.com/openssl/openssl/blob/90c311315c15a4fea895fd317d9c8fe801ba04a0/crypto/sha/sha512.c)), `SHA512_Init` and `SHA512_Update` always succeed. `SHA512_Final` can fail if you pass NULL or the `md` buffer size is insufficient. Both of these conditions are not checked in the original code and can either silently destroy memory or crash.

Let me know how you want to proceed.

File: include/fluent-bit/flb_sha512.h [hunk 1]
```diff
@@ -25,13 +45,26 @@
 
 #include <stdint.h>
 
+#ifdef FLB_HAVE_OPENSSL
```
Comment (leonardo-albertovich) [inline include/fluent-bit/flb_sha512.h:48]: You should include `fluent-bit/flb_info.h` before using this constant, that's a configure time generated header that defines these macros.

**Comment by leonardo-albertovich** [comment]
Once you  fix that one issue I'll be able to merge this.

**Comment by leonardo-albertovich** [comment]
Hey @yuryu, as soon as you're able to add that include line I'll merge this, I'm just pinging you so I don't forget about it.

**Review by leonardo-albertovich — APPROVED** [review]
Top notch!

**Comment by leonardo-albertovich** [comment]
Thanks a lot @yuryu!

**Pull request merged**